apps/docs/refactor.md

```diff
@@ -0,0 +1,565 @@
+- Encapsulation refers to the way in which data and functionality are (or are
+  not) exposed outside a given library. Libraries should expose their data and
+  functionality via a well-defined API that dictates how libraries can interact
+  with each other.
+
+In defining library bounderies, we should strive for a high degree of cohesion
+within a library, little if any implicit coupling with other libraries, and
+strong encapsulation.
+
+Because of the way our product is currently implemented, and because of how web
+technologies work at a fundamental level, there are limitations to how
+encapsulated, cohesive, and uncoupled we can make various libraries. For
+example, any code that renders directly to the DOM and relies on that DOM
+remaining in a particular state, is fundamentally not encapsulated because any
+other code on the page can do whatever it wants with that DOM at any
+time. Libraries like react and redux try to solve this by encouraging specific
+coding patterns that limit the surface area of unencapsulated code, but it is up
+to individual developers to actually stick to those patterns when writing code.
+
+For instances where react and redux do not provide sufficiently well defined
+patterns for our needs, we need to come up with our own patterns that reduce our
+exposure to unencapsulated spaghetti code.
+
+Getting back to "library types", I think there are two different library types
+for which we need to establish well defined patterns.
+
+1. **feature library** - This is a library that provide some meaty feature
+   of a page. It is characterized by the following attributes:
+
+   - maintains UI represented in the DOM (i.e. calls to ReactDOM.render(), or
+     use of jQuery to manipulate the DOM)
+   - interacts with a remote, stateful service (i.e. ajax calls that generate
+     writes to a database)
+   - uses or provides shared application state
+
+   Some examples of "feature libraries" in our code base today would be the code
+   found in `src/applab`, `src/netsim`, and `src/flappy`. Note however that
+   these directories also contain `main.js` files which strictly speaking are
+   not part of the "library" but are actually entry points that get loaded on a
+   particular html page.
+
+2. **capsule library** - This is a library that is wholy encapsulated and
+   decoupled, and which therefore:
+
+   - does not maintain any UI represented in the DOM (react components are OK, but
+     calls to ReactDOM.render() are not)
+   - does not interact with remote, stateful services (ajax calls to read-only
+     services could be OK)
+   - does not use or provide shared application state (at least not in a direct way)
+
+   Some examples of "capsule libraries" in our code base would be
+   `src/MusicController.js`, `src/dom.js`, `src/Sounds.js`.
+
+
+## Capsule Library ##
+
+### Capsule Library Pattern ###
+
+In the case of a relatively simple capsule library that consists of only one
+file, the pattern is pretty simple:
```

Comment:
maybe. I guess the primary goal of the capsule pattern is to make sure you don't accidentally create a feature library :)

Does that work?